python/lsst/cp/pipe/_lookupStaticCalibration.py

```diff
@@ -0,0 +1,74 @@
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from lsst.obs.base import Instrument
+
+__all__ = ["lookupStaticCalibration"]
+
+
+def lookupStaticCalibration(datasetType, registry, quantumDataId, collections):
+    """A lookup function override for QuantumGraph generation that allows a
+    PipelineTask to have an input dataset (usually a camera) that is formally a
+    calibration with a validity range, without having a temporal data ID for
+    the lookup, by asserting that there is in fact only dataset for all time.
+
+    Parameters
+    ----------
+    datasetType : `lsst.daf.butler.DatasetType`
+        Dataset type to look up.
+    registry : `lsst.daf.butler.Registry`
+        Registry for the data repository being searched.
+    quantumDataId : `lsst.daf.butler.DataCoordinate`
+        Data ID for the quantum of the task this dataset will be passed to.
+        This must include an "instrument" key, and should also include any
+        keys that are present in ``datasetType.dimensions``.  If it has an
+        ``exposure`` or ``visit`` key, that's a sign that this function is
+        not actually needed, as those come with the temporal information that
+        would allow a real validity-range lookup.
+    collections : `lsst.daf.butler.registry.CollectionSearch`
+        Collections passed by the user when generating a QuantumGraph.  Ignored
+        by this function (see notes below).
+
+    Returns
+    -------
+    refs : `list` [ `DatasetRef` ]
+        A zero- or single-element list containing the matching dataset, if one
+        was found.
+
+    Notes
+    -----
+    This works by looking in the `~CollectionType.RUN` collection
+    that `lsst.obs.base.Instrument.writeCuratedCalibrations` (currently!) uses,
+    instead of the collections passed into it.  This may be considered
+    surprising by users (but will usually go unnoticed because the dataset
+    returned _is_ actually in those given input colllections, too).  It may
+    stop working entirely once we have data repositories with multiple
+    calibration collections; a better workaround or a more principled change
```

Comment:
This implies that multiple calibration collections are not allowed?  And that `writeCuratedCalibrations` puts those calibrations into a `RUN` collection?  Does that mean they need to be certified into the calibration collection?
Typo on line 5: "colllections".

Reply:
Multiple calibration collections are allowed, but we lack both naming conventions and a good workflow for populating more than one with curated calibrations, and this function essentially relies on those inadequate naming conventions.

More precisely, `writeCuratedCalibrations` first `put`s to multiple `RUN` collections (one for each `calibDate` + one for all unbounded datasets) and then certifies them all into one calibration collection.  It has a suffix parameter for the collection names, to make it possible to write to a totally distinct suite of collections each time it is invoked, but nothing changes the suffix right now (see the comment from Tim that I pinged you on in obs_base).  This workaround will be fine as long `writeCuratedCalibrations` has been invoked with no suffix argument on the repo, and it will do the right thing as long as this is actually the version of the curated calibrations the user wants - it'll pick up the version created with no suffix regardless.

And, of course, it will only work as long as the camera's validity range actually is unbounded.